### PR TITLE
Forward kafka cluster options to clients

### LINKED ISSUE #14123: SASL & SSL Support for kafka config
## Summary

Kafka and Zookeper support encryption and authentication as part of their standard setup, but these configuration options are not available as environment variables in Sentry's KAFKA_CLUSTERS config option or as environment variables to Snuba.

## Motivation

Securing data in transit adds an additional layer of security and compliance for sentry.

## Additional Context

I propose adding  security_protocol, sasl_username, sasl_password and sasl_mechanisms to implementations of KAFKA consumer and producer methods and exposing those options consistently to Sentry, Snuba and related applications.

### PR DESCRIPTION
**Pull request by alexef**
Fixes: #14123. Related to: https://github.com/getsentry/snuba/pull/1406

### CODE REVIEW

File: src/sentry/snuba/query_subscription_consumer.py
Comment (BYK) [inline src/sentry/snuba/query_subscription_consumer.py:72]: This ain' JavaScript my friend, you need to go old fashioned here:

```py
conf = dict({
    "group.id": self.group_id,
    "session.timeout.ms": 6000,
    "session.timeout.ms": 6000,
    "auto.offset.reset": self.initial_offset_reset,
    "auto.offset.reset": self.initial_offset_reset,
    "enable.auto.commit": "false",
    "enable.auto.commit": "false",
    "enable.auto.offset.store": "false",
    "enable.auto.offset.store": "false",
    "enable.partition.eof": "false",
    "enable.partition.eof": "false",
    "default.topic.config": {"auto.offset.reset": self.initial_offset_reset},
    "default.topic.config": {"auto.offset.reset": self.initial_offset_reset},
}, **self.broker_config)
```

or

```py
conf = {
    "group.id": self.group_id,
    "session.timeout.ms": 6000,
    "session.timeout.ms": 6000,
    "auto.offset.reset": self.initial_offset_reset,
    "auto.offset.reset": self.initial_offset_reset,
    "enable.auto.commit": "false",
    "enable.auto.commit": "false",
    "enable.auto.offset.store": "false",
    "enable.auto.offset.store": "false",
    "enable.partition.eof": "false",
    "enable.partition.eof": "false",
    "default.topic.config": {"auto.offset.reset": self.initial_offset_reset},
    "default.topic.config": {"auto.offset.reset": self.initial_offset_reset},
}
conf.update(self.broker_config)
```

File: src/sentry/utils/batching_kafka_consumer.py
Comment (BYK) [inline src/sentry/utils/batching_kafka_consumer.py:218]: Same as above.

File: src/sentry/utils/batching_kafka_consumer.py
Comment (BYK) [inline src/sentry/utils/batching_kafka_consumer.py:232]: Same as above.

**Review by BYK — CHANGES_REQUESTED** [review]
Other than the syntax errors, this seems fine to me.

Deferring to @mattrobenolt and @tkaemming as our resident Kafka experts.

File: src/sentry/snuba/query_subscription_consumer.py
Comment (alexef) [reply src/sentry/snuba/query_subscription_consumer.py:72]: Thanks @BYK! This is actually valid syntax in Python 3. I will refactor the PR to support Python 2.7 as well :)

Comment (BYK) [reply src/sentry/snuba/query_subscription_consumer.py]: @alexef - sorry, I didn't know the `**dict` syntax made it into `dict`s in Py3, that's great and thanks for sharing the info! We are working towards moving on to Python 3.6 but we still need to support Py 2.7 for a while more.

**Review by BYK — APPROVED** [review]
Looks legit to me. Still waiting to hear back from @mattrobenolt and @tkaemming before merging though.

Thanks a lot for looking into this @alexef!

**Review by fpacifici — CHANGES_REQUESTED** [review]
Hi @alexef, happy to see you here as well :D

Thanks for doing this on Sentry as well. 
There are a couple of things to figure out.
1) the order of the override, please see the comment inline.
2) there is a potential backward compatibility issue I would like @BYK to chime in on because it impact self hosted users. Today we are already applying the kafka connection parameters from the settings for producers (https://github.com/getsentry/sentry/blob/master/src/sentry/utils/kafka.py#L34-L35) Which means self hosted users may have already overridden their connection parameters. With this PR, at the next release, those overridden parameters will quietly start being applied to consumers as well. Most likely this is not an issue, but it is impossible to guarantee that nobody overrode a parameter that is fine for producers but breaks everything if applied to their consumers as well. @BYK what do you think? Is this something we have to watch out or communicate in the onpremise installation?

File: src/sentry/eventstream/kafka/consumer.py
Comment (alexef) [reply src/sentry/eventstream/kafka/consumer.py]: Updated.

**Comment by BYK** [comment]
>@BYK what do you think? Is this something we have to watch out or communicate in the onpremise installation?

@fpacifici - good point that said we pre-set this value in on-premise and don't provide support for external Kafka instances. What's the worst-case outcome if anyone modified these settings for their setup?

**Comment by fpacifici** [comment]
@BYK the worst case scenario is very hard to estimate because the amount of possible interactions to rule out between parameters when applied to producers and consumers is huge. It is much less likely that somebody would have overridden a parameter for the producer that actually would not work for the consumer as in the end it is the same broker, but it is not impossible.
I think it would be quite unlikely that a user did an override of the producer config to begin with (today this is not applied to the consumer) they rely on to make things work. Otherwise they would have needed them for the consumers as well. 
So likely we would not run into issues here. 

To be 100% sure we would have to separate producer from consumer config, which is kind of wrong anyway because a lot of it is supposed to be in common.
Also the current situation (before this PR) makes it hard to enforce a list of allowed parameters and prevents us from enforcing a contract with users since we do not enforce what can be changed (to avoid having these issues again).

I would suggest we do the followings:
- define only one config (producer/consumer) in the settings. We can apply specific overrides defined by the product so that what we would see in the config would be only connection related parameters which tend to be common
- Have a list of allowed settings in the product (like I asked @alexef to build in the snuba version (thanks) ) so we can be sure of what can be overridden by settings going on. 
- enforce that list but have an additional setting to disable this behavior so if somebody already added an override to the producer settings they can easily unblock themselves by disabling the enforced config (we would need to communicate this in the next release). Hopefully this behavior is unlikely. Again changing fundamental parameters of the kafka connection on the producer only seems unlikely to me, though it is  not impossible.

@BYK do you think it would be acceptable? At least we would be making it clear and explicit when an unsupported override is happening.

**Comment by BYK** [comment]
@fpacifici your plan sounds like it covers all the cases. It also seems like a longer term plan though. I think:

1. We can land this patch as it is since it is fixing an issue that people are having and the potential fallout from this fix is near zero based on our usage numbers and patterns
2. We can/should implement your suggestion of having a whitelist of options if this is important, but without the enforcement killswitch for the sake of simplicity and speed of implementation, given the case it will cover will be very small

What do you think?

**Comment by fpacifici** [comment]
@byk, unfortunately I found a case where the settings for the producer may not be safe on the consumer that we use internally.
At this point I think the safest option is to split the config between producer and consumer. This would not be too complex for this PR and will also address the backward compatibility issue.

What about we changed the structure of KAFKA_CLUSTERS into
```
KAFKA_CLUSTERS = {
   'default': {
        'common': {
            "bootstrap.servers": "127.0.0.1:9092",
           .....
        },
        "producers": {
             "compression.type": "lz4",
             "message.max.bytes": 50000000,  # 50MB, default is 1MB
        },
       "consumers": {}
       # old parameters here for backward compatibility. These are deprecated and will be removed over time
       "bootstrap.servers": "127.0.0.1:9092",
   }
```

We would do like on snuba, the priority of parameters being loaded wouldl be:
1) hardcoded application overrides 
2) legacy parameters (for backwards compatibility)
3) specific parameters (producer/consumer)
4) common parameters

@alexef What do you think about this ?

**Comment by alexef** [comment]
Updated. I couldn't find any code overrides for configs, this is why I didn't add it to the config getter.

@fpacifici let me know what you think

File: src/sentry/utils/kafka.py
Comment (fpacifici) [inline src/sentry/utils/kafka.py:104]: Legacy options should only apply to the producer (as it is now).

File: src/sentry/utils/kafka.py
Comment (fpacifici) [inline src/sentry/utils/kafka.py:98]: As in snuba, please remove this, it can generate a lot of logs. We will add it back after communicating that this config is deprecated.

File: src/sentry/utils/batching_kafka_consumer.py
Comment (fpacifici) [inline src/sentry/utils/batching_kafka_consumer.py:217]: I am a bit uncomfortable with these `copy()` spread around where we need to override the config. 
Is there any issue with calling `get_kafka_consumer_cluster_options` here and passing the dictionary of overrides so that that function takes care of doing the copies when needed in a transparent way ?

**Review by fpacifici — CHANGES_REQUESTED** [review]
Thanks, this goes in the right direction.
A few comments inline. 
Could you also write a short test for your config loader methods? This does not need to be as exhaustive as the one you built for snuba.

File: src/sentry/utils/batching_kafka_consumer.py
Comment (alexef) [reply src/sentry/utils/batching_kafka_consumer.py]: Refactored. Did my best to avoid .copy() calls - had to restructure a bit the Synchronized consumer and tests

**Comment by alexef** [comment]
Up for another check :)

File: src/sentry/snuba/query_subscription_consumer.py
Comment (fpacifici) [inline src/sentry/snuba/query_subscription_consumer.py:71]: @wedamija, do these properties (self.group_id and self.initial_offset_reset) ever get modified on an instance of this class? Before this PR the config dictionary was created every time we executed `run`. Now this is created only once in the constructor, so if those attributes are mutable this may be an issue.

File: src/sentry/utils/kafka_config.py
Comment (fpacifici) [inline src/sentry/utils/kafka_config.py:21]: I think the order these parameters are being applied is not the right one.
I would say it should be:
- common_options
- custom_options
OR 
- legacy options (if present)
THEN override_options.

So legacy options replace the combination of custom and common options. Legacy should not exist if somebody populates the legacy options we use those and we do not make them interact with the new config version. Merging legacy and new could give unpredictable results, at least this way it would be more predictable.

Then override are applied on top of this since they are chosen and forced by the application in code (which means they should never be overridden).

File: src/sentry/utils/kafka_config.py
Comment (fpacifici) [inline src/sentry/utils/kafka_config.py:22]: Could you please do like for snuba and have a list of permitted parameters (that applies only to common and custom options, since legacy is not something we can control) ?

File: src/sentry/utils/kafka.py
Comment (fpacifici) [inline src/sentry/utils/kafka.py:69]: This means that the `bootstrap_servers` parameter (between legacy options) is supposed to be accepted whether it is a string with comma separated brokers or whether it is a sequence. Thus, sadly, you will have to allow for the same behavior in your kafka_config methods to keep this backward compatible. 
So the kafka consumer wants a string, if the config contains a sequence you need to join it and if it is a string I think you need to remove the commas.
I think you can get away in only supporting this for the legacy parameter, let's force common and custom parameters to be a sequence.

**Review by fpacifici — CHANGES_REQUESTED** [review]
Thanks.
A few items that need to be corrected.
Also could you please add a unit test for your new functions in kafka_config ? It does not need to be as thorough as the one in snuba, but still the basic test coverage we already have is not doing a good job for those corner cases specifically since you would never have legacy parameters anywhere in the existing tests.

File: src/sentry/utils/kafka.py
Comment (alexef) [reply src/sentry/utils/kafka.py:69]: In snuba we use comma separated strings, as this is also what librdkafka expects.
I will update common and custom to expect strings as well, not a sequence.

File: src/sentry/snuba/query_subscription_consumer.py
Comment (alexef) [reply src/sentry/snuba/query_subscription_consumer.py:71]: I've refactored with confidence as I saw that the only place this class is instantiated is within a click cli command.
But I might have missed access to these fields, if they are mutable.

**Comment by BYK** [comment]
@fpacifici - sorry, this got buried under other things. Your suggestion is mostly good for me but I'd propose dropping the `common` top-level key, keeping the old variables and adding `producer` and `consumer` as top-level keys.

The possibility to clash with a real kafka config for those is very small (0%?) and if you think that's an issue we can call them `__producer` and `__consumer`.

This would remove the pain of deprecation and config migration and also make the config a bit simpler.

Thoughts?

**Comment by fpacifici** [comment]
> This would remove the pain of deprecation and config migration and also make the config a bit simpler.

Removing the `common` top level key does not remove migration or deprecation. That is used just for convenience not two duplicate settings that do not need to be duplicated (like the host). 
The deprecation, and migration only concerns the top level individual keys (outside `common`, `producer` and `consumer`) which we cannot really remove right away because of immediate backward compatibility issue (even with our production environment).

**Comment by BYK** [comment]
>Removing the common top level key does not remove migration or deprecation. That is used just for convenience not two duplicate settings that do not need to be duplicated (like the host).

This is what I'm saying though. Can we omit the `common` namespace and dump all common variables at the top level?

**Comment by fpacifici** [comment]
> Can we omit the common namespace and dump all common variables at the top level?

Not exactly. Right now those legacy settings are onnly applied to the producer and not to the consumer (we do it also in our environments). If we started treating those parameters as common we would start applying them to consumers as well thus potentially causing issues (it is likely to happen for our prod as well).

**Comment by alexef** [comment]
@fpacifici I've updated the PR with the requested unittest. Please have a look again :)

File: src/sentry/utils/kafka_config.py
Comment (fpacifici) [inline src/sentry/utils/kafka_config.py:55]: Up to you,
I would not block the PR for this: I think it would me more intuitive (for the new configuration, where this is possible) to have the bootstrap servers in the config represented as a sequence (harder to mess up the structure), and then turn it into a string (here in this module) instead of doing the opposite. This would follow the same convention we have in snuba.

Unfortunately we still need to support the strings for the legacy config, but that, being legacy will be eventually removed, so we can eventually get to a consistent representation.

**Review by fpacifici — COMMENTED** [review]
Sorry for the delay.
You have a failure in your test. (you can run it individually by simply running `pytest YOUR_TEST_PATH` from the sentry root directory when in the sentry virtual ev.

**Comment by alexef** [comment]
Ah, I forgot to push the fixed test. Yeah, I messed up my local environment, this is why I couldn't run tests locally.

There is an edge case that this broken test revelead: we don't use legacy `bootstrap.servers` when initializing a consumer client. I believe this is wrong, as (same as for snuba) existing installation could have overridden the bootstrap.servers settings.

Let me push another change that preservers legacy bootstrap.servers (if present) for consumer.

Re: bootstrap.servers, the way it is currently implemented in this PR _is consistent_ with snuba. There we also use strings, not lists: https://github.com/getsentry/snuba/blob/master/snuba/settings.py#L79

File: src/sentry/utils/kafka_config.py
Comment (alexef) [reply src/sentry/utils/kafka_config.py:55]: at this stage in code, the bootstrap.servers must be a comma separated string, as this is what librdkafka is expected. 
sure we can make it a list before (same as it was in the legacy config prior these changes)

File: src/sentry/utils/kafka_config.py
Comment (fpacifici) [inline src/sentry/utils/kafka_config.py:48]: Thanks for noticing this, but I don't think this is the right fix. If there is a bootstrap_server parameter I think we should ignore the custom fields just like we do in the branch above. 
Not doing so means we would have the consumer and the producer configuration behave differently: For producers, the mere presence of legacy fields will make us ignore new fields. For consumers, we would always pick the new field and override a specific one with the legacy config.

I would make them behave the same. So if there is a bootstrap.servers legacy field ignore the new fields in the consumer as well.

**Comment by fpacifici** [comment]
Meanwhile I will try to figure out what is wrong with the failing test: onpremise-github-pr-check

File: src/sentry/utils/kafka_config.py
Comment (alexef) [reply src/sentry/utils/kafka_config.py:48]: you are right again 🤦 . pushing the fix

Comment (alexef) [reply src/sentry/utils/kafka_config.py:48]: updated

**Comment by fpacifici** [comment]
Could you please merge master in your branch, I don't think there should be a reason for the gcb build (onpremise-github-pr-check) to fail when I trigger it.

**Comment by alexef** [comment]
Merged master. I don't get why https://github.com/getsentry/sentry/pull/21275/checks?check_run_id=1377976970 started failing :-/

**Comment by fpacifici** [comment]
Looking into the `js build and lint / typescript and lint` it is certainly not related with your PR.

**Comment by fpacifici** [comment]
Apparently `onpremise-github-pr-check` is actually failing only on this PR, so we need to figure out what the reason is. Will try again tomorrow.

**Comment by fpacifici** [comment]
It seems `onpremise-github-pr-check` is a actual issue with this PR (tried with others from forks and could not reproduce). Looking into why it fails but if you want to try as well, this is the step that fails:
https://github.com/getsentry/sentry/blob/master/docker/cloudbuild.yaml#L60-L82. Can you access the cloudbuild results? If not it seems coming from here:

```
192.168.10.2 - - [10/Nov/2020:20:21:39 +0000] "GET /api/0/projects/sentry/internal/events/c505c97773a630f11451d85215104ae3/ HTTP/1.1" 404 28 "http://nginx" "curl/7.58.0" "-"
relay_1                        | 2020-11-10T20:21:12Z [relay_server::actors::upstream] ERROR: authentication encountered error: could not send request to upstream
relay_1                        |   caused by: Failed to connect to host: Connection refused (os error 111)
relay_1                        |   caused by: Connection refused (os error 111)
relay_1                        |   caused by: Connection refused (os error 111)
relay_1                        | 2020-11-10T20:21:12Z [relay_server::actors::upstream] ERROR: authentication encountered error: could not send request to upstream
relay_1                        |   caused by: Failed to connect to host: Connection refused (os error 111)
relay_1                        |   caused by: Connection refused (os error 111)
relay_1                        |   caused by: Connection refused (os error 111)
relay_1                        | 2020-11-10T20:21:13Z [relay_server::actors::upstream] ERROR: authentication encountered error: could not send request to upstream
relay_1                        |   caused by: Failed to connect to host: Connection refused (os error 111)
relay_1                        |   caused by: Connection refused (os error 111)
relay_1                        |   caused by: Connection refused (os error 111)
relay_1                        | 2020-11-10T20:21:20Z [relay_server::actors::upstream] ERROR: authentication encountered error: could not send request to upstream
relay_1                        |   caused by: Timeout while waiting for response
```

It seems that relay is not capable to send the ingested event to sentry in the end to end test. So I need to dig into that to reproduce.

**Comment by fpacifici** [comment]
While https://github.com/getsentry/sentry/pull/21893 should fix the js build and lint step that broke for forks two days ago.

**Comment by fpacifici** [comment]
`js build and lint / typescript and lint` should be fixed, please merge master.

**Comment by alexef** [comment]
> Can you access the cloudbuild results? If not it seems coming from here:

@fpacifici Unfortunately I cannot access the google cloud build results. I try to look into the code changes for relay to see if there is anything obvious.

I'm now merging and fixing conflicts with master.

File: src/sentry/snuba/query_subscription_consumer.py
Comment (alexef) [reply src/sentry/snuba/query_subscription_consumer.py:73]: ping @wedamija :)

**Comment by fpacifici** [comment]
I found the issue that causes the cloudbuild to fail:

```
07:58:38 [ERROR] celery.app.trace: Task sentry.tasks.store.save_event[6c0d8572-1805-4c7a-ab2d-e5bbb37348aa] raised unexpected: 
ValueError('The `socket.timeout.ms` configuration key is not supported.',) (data={u'internal': False, u'traceback': u'Traceback (most recent call last):
File "/usr/local/lib/python2.7/site-packages/celery/app/trace.py", line 412, in trace_task  
    R = retval = fun(*args, **kwargs)
File "/usr/local/lib/python2.7/site-packages/celery/app/trace.py", line 704, in __protected_call__
    return self.run(*args, **kwargs)
File "/usr/local/lib/python2.7/site-packages/sentry_sdk/integrations/celery.py", line 186, in _inner
    reraise(*exc_info)
File "/usr/local/lib/python2.7/site-packages/sentry_sdk/integrations/celery.py", line 181, in _inner
    return f(*args, **kwargs)
File "/usr/local/lib/python2.7/site-packages/sentry/tasks/base.py", line 48, in _wrapped
    result = func(*args, **kwargs)
File "/usr/local/lib/python2.7/site-packages/sentry/tasks/store.py", line 870, in save_event
    _do_save_event(cache_key, data, start_time, event_id, project_id, **kwargs)
File "/usr/local/lib/python2.7/site-packages/sentry/tasks/store.py", line 784, in _do_save_event
    project_id, assume_normalized=True, start_time=start_time, cache_key=cache_key
File "/usr/local/lib/python2.7/site-packages/sentry/utils/metrics.py", line 193, in inner
    return f(*args, **kwargs)
File "/usr/local/lib/python2.7/site-packages/sentry/event_manager.py", line 478, in save
    _eventstream_insert_many(jobs)
File "/usr/local/lib/python2.7/site-packages/sentry/utils/metrics.py", line 193, in inner
    return f(*args, **kwargs)
File "/usr/local/lib/python2.7/site-packages/sentry/event_manager.py", line 768, in _eventstream_insert_many
    skip_consume=job.get("raw", False),
File "/usr/local/lib/python2.7/site-packages/sentry/utils/services.py", line 105, in <lambda>
    context[key] = (lambda f: lambda *a, **k: getattr(self, f)(*a, **k))(key)
File "/usr/local/lib/python2.7/site-packages/sentry/eventstream/snuba.py", line 132, in insert
    headers={"Received-Timestamp": six.text_type(received_timestamp)},
File "/usr/local/lib/python2.7/site-packages/sentry/eventstream/kafka/backend.py", line 52, in _send
    self.producer.poll(0.0)
File "/usr/local/lib/python2.7/site-packages/django/utils/functional.py", line 35, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
File "/usr/local/lib/python2.7/site-packages/sentry/eventstream/kafka/backend.py", line 25, in producer
    return kafka.producers.get(settings.KAFKA_EVENTS)
File "/usr/local/lib/python2.7/site-packages/sentry/utils/kafka.py", line 36, in get
    cluster_options = get_kafka_producer_cluster_options(cluster_name)
File "/usr/local/lib/python2.7/site-packages/sentry/utils/kafka_config.py", line 65, in get_kafka_producer_cluster_options
    return _get_kafka_cluster_options(cluster_name, PRODUCERS_SECTION, with_legacy=True)
File "/usr/local/lib/python2.7/site-packages/sentry/utils/kafka_config.py", line 54, in _get_kafka_cluster_options
    configuration_key=configuration_key\nValueError: The `socket.timeout.ms` configuration key is not supported.\n', u'name': 'sentry.tasks.store.save_event', u'args': '()', u'kwargs': "{'event_id': '87921e4de05e79d26f28ff33660dca96', 'project_id': 1L, 'start_time': 1605254317.0, 'cache_key': u'e:87921e4de05e79d26f28ff33660dca96:1', 'data': None}", u'description': u'raised unexpected', u'hostname': '7b4af7b6291f', u'id': '6c0d8572-1805-4c7a-ab2d-e5bbb37348aa', u'exc': "ValueError('The `socket.timeout.ms` configuration key is not supported.',)"})
```

Thats' because the PR applies the list of allowed parameters not only to the new configuration (correct) but also to the legacy configuration (incorrect). The legacy configuration should not be validated since we do not have control on what was added so far.
Also `socket.timeout.ms` is something that should be in the allowed list.

This issue shows up on cloud build (fortunately it got caught there) because it runs the onpremise version of sentry and that one overrides that parameter by default: https://github.com/getsentry/onpremise/blob/master/sentry/sentry.conf.example.py#L119

If you want to reproduce this:
1) install self hosted sentry as explained in https://github.com/getsentry/onpremise
2) execute the tests `./tests.sh` It should work.
3) replace the docker image for sentry with the one fromo your cloud build in this file: https://github.com/getsentry/onpremise/blob/master/.env#L6. That has to be replaced with `us.gcr.io/sentryio/sentry:00f4a3978a6ed00cb0a4ea82b343422129976f54`. I picked it from the logs of your build, but you cann technically build the image locally if you prefer.
4) find the logs of the celery worker that is the one that throws (which is not shown in the logs of the cloud build): `docker log sentry_onpremise_worker_1`

@BYK The celery worker logs are not shown in cloudbuild upon failure, that means every errors happening in sentry during ingestion is hidden. That covers all the steps from the ingestion consumer (after relay) and before eventstream (where we produce to kafka for snuba)

**Comment by BYK** [comment]
Quite sure the failure is due to this config:

https://github.com/getsentry/onpremise/blob/066bf262aac7b50cdc870b3f4e41d378cc2193db/sentry/sentry.conf.example.py#L116-L125

I think if we cannot make this backward-compatible with these existing configurations out in the wild, we should rename this config.

**Comment by BYK** [comment]
Not sure if we can do something clever here to auto-migration/compatibility: https://github.com/getsentry/sentry/blob/master/docker/sentry.conf.py

**Comment by fpacifici** [comment]
@BYK, let me clarify. There is nothing wrong with `socket.timeout.ms`.  This PR is supposed to be backward compatible and support it without changes to the field, though right now it is applying the same validation to the legacy fields that it applies to the new ones. This is incorrect and @alexef should be able to fix that behavior quickly.

I tagged you because of the missing logs in the cloudbuild results. I think we should expose the celery worker logs as well, because there is where a big chunk of the ingestion logic in the sentry code base happens.

**Comment by BYK** [comment]
@fpacifici - sorry, I wrote my comment before seeing yours. I get the issue now. I've also submitted a PR to make the logs available.

**Comment by fpacifici** [comment]
thanks

File: src/sentry/utils/kafka_config.py
Comment (fpacifici) [inline src/sentry/utils/kafka_config.py:46]: This is going to be very hard to grasp. We need to change how this is structured.
Basically what you are trying to do is to pick all legacy options for producers (if there are legacy options) and, for consumers, pick only bootstrap.servers from legacy options.
Why not rewriting it as
```
#Rename with_legacy into only_bootstrap and invert
if legacy_options:
   if only_bootstrap:
       assert "bootstrap.servers" in legacy_options
       options["bootstrap.servers"] = legacy_options["bootstrap.servers"]
   else :
       options.update(legacy_options)
else:
   # same as before.
```

So it should be clear that picking the bootstrap only from the legacy option is not a way to override other options but it is meant to build legacy options that discard everything that is not legacy.

**Review by fpacifici — CHANGES_REQUESTED** [review]
Just one small issue in the kafka_config, but that is fairly important, it took me a while to figure out why it was working in a certain way. Then everything else pass, tomorrow I will merge if it is fixed.

**Comment by BYK** [comment]
Thanks a lot @alexef for following along for all this time!

File: src/sentry/utils/kafka_config.py
Comment (alexef) [reply src/sentry/utils/kafka_config.py]: yup, your version is way more readable. this ended up like this due to many rewrittings.

updated

**Comment by alexef** [comment]
@BYK I can't wait to see this merged so we can upgrade our onprem installation. 

The biggest challenge was keeping my dev environment usable - python2.7 + mac does not play well, but this should be addressed soon.

**Review by fpacifici — APPROVED** [review]
Thank you, this will be very useful.

**Pull request merged**
